# Data Cleaning with Pandas

In this notebook we'll go through a few basic data cleaning steps that should be performed on all new datasets where necessary.

We'll go through the process with both the `orders` and `orderlines` datasets. You can then practice these skills by cleaning the `products` dataset yourself

In [1]:
import pandas as pd

In [2]:
# orders.csv
url = "https://drive.google.com/file/d/1Vu0q91qZw6lqhIqbjoXYvYAQTmVHh6uZ/view?usp=sharing"
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
orders = pd.read_csv(path)

# orderlines.csv
url = "https://drive.google.com/file/d/1FYhN_2AzTBFuWcfHaRuKcuCE6CWXsWtG/view?usp=sharing"
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
orderlines = pd.read_csv(path)

Before we begin, let's create a copy of the `orders` and `orderlines` DataFrames. This way we are sure any of our changes won't affect the original DataFrames

In [3]:
orders_df = orders.copy()

orderlines_df = orderlines.copy()

One of the best ways to begin data cleaning is by exploring using `.info()`. This will tell us:
* The shape of the DataFrame
* The names of the columns
* If there are any missing values
* The datatypes of the columns

By exploring the missing values and correcting any incorrect datatypes, we often come across inconsistencies in our data.

Beyond this, we should also have a **check for any duplicate rows**.

Let's first deal with the duplicates, as it's nice and easy, then we'll explore what `.info()` has to tell us.

## 1.&nbsp; Duplicates
We can check for duplicates using the pandas [.duplicated()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.duplicated.html) method.

We can then delete these rows, if we wish, using [.drop_duplicates()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html)

In [30]:
# orders_df
orders_df.duplicated().sum()

0

In [8]:
# orderlines_df
orderlines_df.duplicated().sum()

0

We have no duplicate rows in either DataFrame. Easy, there is no problem to solve. Normally though, if there were some duplicates, we'd drop the extra rows.

# 2.&nbsp; `.info()`

In [9]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226909 non-null  int64  
 1   created_date  226909 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226909 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 6.9+ MB


* `total_paid` has 5 missing values
* `created_date` should become datetime datatype

In [10]:
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


* `date` should be a datetime datatype
* `unit_price` should be a float datatype

## 3.&nbsp; Missing values

### 3.1.&nbsp; Orders
* `total_paid` has 5 missing values

In [14]:
print(f"5 missing values represents {((orders_df.total_paid.isna().sum() / orders_df.shape[0])*100).round(5)}% of the rows in our DataFrame")

5 missing values represents 0.0022% of the rows in our DataFrame


In [17]:
orders_df.shape[0]

226909

In [18]:
orders_df.total_paid.isna().sum()

5

A quick way to find out a percentage here, if you don't need to print out a sentence for yourself/students/colleagues is `.value_count(normalize=True)`

In [23]:
orders_df.total_paid.isna().value_counts(normalize=True)

False    0.999978
True     0.000022
Name: total_paid, dtype: float64

As there is such a tiny amount of missing values, we will simply delete these rows, as we have enough data without them.

In [24]:
orders_df = orders_df.loc[~orders.total_paid.isna(), :]

In [33]:
orders_df.total_paid.isna().sum()

0

In [34]:
orders_df.total_paid.notna().sum()

226904

In [25]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 226904 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226904 non-null  int64  
 1   created_date  226904 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226904 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 8.7+ MB


Should you have a significant number of missing values in the future, you have a choice:
+ you can impute the values
+ you can take the values from other DataFrames, if they are present there
+ you can delete the values
+ or any number of other creative solutions

Please, always consider how much time you have on your project, and what impact your method of choice will have on your final assesment.

### 3.2.&nbsp; Orderlines
There are no missing values in `orderlines_df`

## 4.&nbsp; Datatypes

### 4.1.&nbsp; Orders
* `created_date` should become datetime datatype

In [26]:
orders_df["created_date"] = pd.to_datetime(orders_df["created_date"])

In [35]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 226904 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      226904 non-null  int64         
 1   created_date  226904 non-null  datetime64[ns]
 2   total_paid    226904 non-null  float64       
 3   state         226904 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(1)
memory usage: 8.7+ MB


### 4.1.&nbsp; Orderlines
* `date` should be a datetime datatype
* `unit_price` should be a float datatype

#### 4.1.1.&nbsp; `date`

In [27]:
orderlines_df["date"] = pd.to_datetime(orderlines_df["date"])

In [36]:
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                293983 non-null  int64         
 1   id_order          293983 non-null  int64         
 2   product_id        293983 non-null  int64         
 3   product_quantity  293983 non-null  int64         
 4   sku               293983 non-null  object        
 5   unit_price        293983 non-null  object        
 6   date              293983 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(4), object(2)
memory usage: 15.7+ MB


#### 4.1.2.&nbsp;`unit_price`

In [37]:
orderlines_df["unit_price"] = pd.to_numeric(orderlines_df["unit_price"])

ValueError: Unable to parse string "1.137.99" at position 6

As you can see when we try to convert `unit_price` to a numerical datatype, we receive a `ValueError` telling us that pandas doesn't understand the number `1.137.99`. This is probably because numbers cannot have 2 decimal points. Let's see if there are any other numbers like this.

In [38]:
orderlines_df.unit_price.str.contains("\d+\.\d+\.\d+").value_counts()

False    257814
True      36169
Name: unit_price, dtype: int64

Looks like over 36000 rows in `orderlines` are affected by this problem. Let's work out how much that is as a percentage of our total data.

In [39]:
two_dot_percentage = ((orderlines_df.unit_price.str.contains("\d+\.\d+\.\d+").value_counts()[1] / orderlines_df.shape[0])*100).round(2)
print(f"The 2 dot problem represents {two_dot_percentage}% of the rows in our DataFrame")

The 2 dot problem represents 12.3% of the rows in our DataFrame


This is a bit of a tricky decision as 12.3% is a significant amount of our data... and we might even end up losing a larger portion of our data than this too. For the moment we will delete the rows as we only have 2 weeks for this project and I'd like some quick, accurate results to show. If we have time at the end, we can come back and investigate this problem further, maybe there's a solution?

Each row of `orderlines` represents a product in an order. For example, if order number 175 contained 3 seperate products, then order 175 would have 3 rows in `orderlines`, one row for each of the products. If 2 of those products have 'normal' prices (14.99, 15.85) and 1 has a price with 2 decimal points (1.137.99), we need to remove the whole order and not just the affected row. If we only remove the row with 2 decimal places then any later analysis about products and prices could be misleading.

We therefore need to find the order numbers associated with the rows that have 2 decimal points, and then remove all the associated rows.

In [40]:
two_dot_order_ids_list = orderlines_df.loc[orderlines_df.unit_price.str.contains("\d+\.\d+\.\d+"), "id_order"]
orderlines_df = orderlines_df.loc[~orderlines_df.id_order.isin(two_dot_order_ids_list)]

In [41]:
orderlines_df.shape[0]

216250

We still have 216250 rows in orderlines to work with. This should be more than enough for our evaluation.

Now that all of the 2 decimal point prices have been removed, let's try again to convert the column `unit_price` to the correct datatype.

In [42]:
orderlines_df["unit_price"] = pd.to_numeric(orderlines_df["unit_price"])

In [43]:
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 216250 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                216250 non-null  int64         
 1   id_order          216250 non-null  int64         
 2   product_id        216250 non-null  int64         
 3   product_quantity  216250 non-null  int64         
 4   sku               216250 non-null  object        
 5   unit_price        216250 non-null  float64       
 6   date              216250 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(4), object(1)
memory usage: 13.2+ MB


It worked perfectly

# Challenge: Clean the `products` DataFrame
Now it's your turn. Use the lessons you learnt above and clean the products DataFrame. You don't have to copy exactly what we did. Think about the consequences of your actions, sometimes it is ok to delete rows, other times you may wish to come up with more creative solutions.

In [48]:
# products.csv
url = "https://drive.google.com/file/d/1afxwDXfl-7cQ_qLwyDitfcCx3u7WMvkU/view?usp=sharing"
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
products = pd.read_csv(path)
products.head(15)

sku                                               name  \
0     RAI0007                  Silver Rain Design mStand Support   
1     APP0023                  Apple Mac Keyboard Keypad Spanish   
2     APP0025                   Mighty Mouse Apple Mouse for Mac   
3     APP0072      Apple Dock to USB Cable iPhone and iPod white   
4     KIN0007        Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
5     APP0073     Apple Composite AV Cable iPhone and iPod white   
6     KIN0008        Mac Memory Kingston 1GB 667MHz DDR2 SO-DIMM   
7     KIN0009        Mac Memory Kingston 2GB 800MHz DDR2 SO-DIMM   
8   KIN0001-2  Mac memory Kingston 4GB (2x2GB) 667MHz DDR2 SO...   
9     APP0100             Apple Adapter Mini Display Port to VGA   
10    APP0101              Apple Mini DisplayPort to DVI Adapter   
11    SEN0021  Sennheiser CX 300-II Precision headphones iPho...   
12    APP0111                     Apple USB Ethernet Adapter Mac   
13    APP0156              45W MagSafe charger Apple MacBook Air   
14    MOS0013         Adapter Moshi FireWire 400 to FireWire 800   

                                                 desc  price promo_price  \
0        Aluminum support compatible with all MacBook  59.99     499.899   
1           USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                              mouse Apple USB cable.     59     569.898   
3               IPhone dock and USB Cable Apple iPod.     25     229.997   
4   2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   
5   IPhone and iPod AV Cable Dock to Composite Video.     45     420.003   
6   1GB RAM Mac mini and iMac (2006/07) MacBook Pr...  18.99     146.471   
7        2GB RAM iMac with Intel Core 2 Duo (Penryn).  36.99     274.694   
8   RAM 4GB (2x2GB) Mac mini and iMac (2006/07) Ma...     74     669.904   
9   Adapter Mini Display Port to VGA MacBook and M...     35     330.003   
10  Adapter mini DisplayPort to DVI MacBook and Ma...     35     330.003   
11     Headphones iPhone iPad iPad 2 iPad 3 and iPod.  49.99     449.878   
12                       USB to Ethernet adapter Mac.     35     330.003   
13     Charger MacBook Air MagSafe 45 W (MC747Z / A).     89     849.904   
14           FireWire 400 adapter cable FireWire 800.     20     179.891   

    in_stock      type  
0          1      8696  
1          0  13855401  
2          0      1387  
3          0      1230  
4          1      1364  
5          0      1230  
6          0      1364  
7          0      1364  
8          0      1364  
9          0      1325  
10         0      1325  
11         0      5384  
12         0      1334  
13         0  13005399  
14         0      1325

In [46]:
# your code here
products_df=products.copy()

In [47]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19326 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          19326 non-null  object
 1   name         19326 non-null  object
 2   desc         19319 non-null  object
 3   price        19280 non-null  object
 4   promo_price  19326 non-null  object
 5   in_stock     19326 non-null  int64 
 6   type         19276 non-null  object
dtypes: int64(1), object(6)
memory usage: 1.0+ MB


# Data Type Errors
 products_df problems found: price data type object, should be float. promo_price data type object, should be float, type data type should be int, but object.


In [50]:
products_df.shape

(19326, 7)

### Look for Duplicates

In [51]:
# your code here
products_df.duplicated().sum()

8746

there are 8746 duplicated values

In [57]:
len(products_df)

19326

before removing duplicates , there are 19326 rows in products_df dataframe

In [58]:
products_df.drop_duplicates(inplace=True)

In [59]:
products_df.shape

(10580, 7)

after removing duplicates, there are 10580 rows in products_df dataframe




# Missing values

1.   desc has missing values of 7
2.   price has missing values of 46

3.   type has missing values of 50







### Look for Missing values


In [60]:
products_df.head()

sku                                           name  \
0  RAI0007              Silver Rain Design mStand Support   
1  APP0023              Apple Mac Keyboard Keypad Spanish   
2  APP0025               Mighty Mouse Apple Mouse for Mac   
3  APP0072  Apple Dock to USB Cable iPhone and iPod white   
4  KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   

                                                desc  price promo_price  \
0       Aluminum support compatible with all MacBook  59.99     499.899   
1          USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                             mouse Apple USB cable.     59     569.898   
3              IPhone dock and USB Cable Apple iPod.     25     229.997   
4  2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   

   in_stock      type  
0         1      8696  
1         0  13855401  
2         0      1387  
3         0      1230  
4         1      1364

In [62]:
products_df.desc.isna().sum() # 7 in desc column

7

In [63]:
products_df.price.isna().sum() #46 missing in price column

46

In [65]:
products_df.type.isna().sum() #50 missing in type column

50

####Missing value in price
checking for products when price has missing values, one pattern can be found: 45 of them has 0 in_stock, their name contains ***Celly*** and SKU starts with **CEL**

In [75]:
products_df[products_df.price.isna()]

sku                                               name  \
34       TWS0019  Twelve South MagicWand support Apple Magic Tra...   
1900     AII0008           Aiino Case MacBook Air 11 '' Transparent   
2039     CEL0020  Celly Ambo Luxury Leather Case + iPhone 6 Case...   
2042     CEL0007  Celly Wallet Case with removable cover Black i...   
2043     CEL0012            Celly Silicone Hard Shell iPhone 6 Blue   
2044     CEL0014        Celly Silicone Hard Shell iPhone 6 Amarillo   
2049     CEL0015    Celly fur-lined Powerbank battery 4000mAh Black   
2051     CEL0018     Celly Wallet Leather Case cover Black iPhone 6   
2052     CEL0023  Celly Ambo Luxury Leather Case + Case Gold iPh...   
2053     CEL0025  Celly Ambo Luxury Leather Case + Case iPhone 6...   
8000     APP1197  Apple iMac 21.5 "Core i5 31 GHz Retina display...   
12052    CEL0028  Celly Swing Arm button selfie with Bluetooth B...   
12053    CEL0029  Celly extensible arm selfie with Bluetooth but...   
12054    CEL0030  Celly Smart Drive Housing + clip car holder iP...   
12055    CEL0031  Celly Smart Drive Case + Car clip support iPho...   
12059    CEL0032      Celly Gost clamp fixing car + Magnetica Black   
12060    CEL0042  Celly iPhone Case Fluor 6s / 6 Black / Transpa...   
12061    CEL0043  Celly Fluor Case iPhone 6 / 6S Pink / Transparent   
12062    CEL0044  Celly iPhone Case Fluor 6s / 6 Yellow / Transp...   
12064    CEL0038           Celly Frost Cover for iPhone 6 / 6S Rosa   
12067    CEL0033  Celly Waterproof iPad Case / Air / Pro 97 "Tra...   
12068    CEL0034  Celly Ambo cover with lid iPhone 6 / 6S Plus B...   
12069    CEL0035  Celly Ambo cover with lid iPhone 6 / 6S Plus B...   
12070    CEL0036  Celly Adapter Micro USB connector to Lightning...   
12071    CEL0051   Celly 4000mAh Battery Power Bank Aluminum Dorado   
12072    CEL0052   Celly 4000mAh Battery Power Bank Aluminum Silver   
12073    CEL0053             Celly IceCube Case iPhone 6 / 6S Black   
12074    CEL0054              Celly IceCube Case iPhone 6 / 6S Rosa   
12075    CEL0055             Celly IceCube Case iPhone 6 / 6S White   
12076    CEL0056          IceCube Celly iPhone Case 6 / 6s Amarillo   
12077    CEL0057        Celly IceCube Case iPhone 6 / 6S Plus Black   
12078    CEL0058         Celly IceCube Case iPhone 6 / 6S Plus Rosa   
12079    CEL0059        Celly IceCube Case iPhone 6 / 6S Plus White   
12080    CEL0060     Celly IceCube Case iPhone 6 / 6S Plus Amarillo   
12088    CEL0040     Celly Frost Cover for iPhone 6 / 6S Plus White   
12089    CEL0041              Celly iPhone 6S Plus Frost Case Black   
12092    CEL0045         Laser Case Celly iPhone 6 / 6S Dark Silver   
12093    CEL0046        Laser Case Celly iPhone 6 / 6S Clear Silver   
12095    CEL0047              Laser Case Celly iPhone 6 / 6S Dorado   
12096    CEL0048                Laser Case Celly iPhone 6 / 6S Rosa   
12097    CEL0049         Laser Case Celly iPhone 6 / 6S Plus Dorado   
12098    CEL0050   Laser Case Celly iPhone 6 / 6S Plus Silver Clear   
12100    CEL0037          Celly Frost Cover for iPhone 6 / 6S Black   
12238    CEL0027          Celly Frost Cover for iPhone 6 / 6S White   
12242    CEL0039       Celly Frost Cover for iPhone 6 / 6S Amarillo   
12947  CEL0034-A    (Open) Celly iPhone 6S Plus Housing Frost White   

                                                    desc price promo_price  \
34     MagicWand for wireless keyboard and Magic Trac...   NaN     299.899   
1900       MacBook Air 11-inch casing with matte finish.   NaN       22.99   
2039   Cover and housing together with magnet for iPh...   NaN     399.905   
2042              Case Book for iPhone 6 card case type.   NaN     128.998   
2043                       Hard Shell Silicone iPhone 6.   NaN        4.99   
2044                       Hard Shell Silicone iPhone 6.   NaN      59.895   
2049   Leather-wrapped External Battery 4000mAh for i...   NaN     239.895   
2051   Card case with transparent protective cover 



####missing value in type


Looking for products when ***type*** has missing values, majority of them this pattern **SKU** starts with APP, **name** has Apple iphone, **desc** has new iPhone, and **promo_price** has two decimal point data

In [74]:
products_df[products_df.type.isna()]

sku                                               name  \
307      SAN0017       SanDisk Cruzer Edge USB 2.0 Flash Drive 16GB   
530      SAN0026    SanDisk Extreme Cruzer 16GB USB 3.0 Flash Drive   
792      APP0672                    Apple iPhone 5S 16GB Space Gray   
797      APP0673                        Apple iPhone 5S 16GB Silver   
798      APP0675                    Apple iPhone 5S 32GB Space Gray   
1193     APP0823                         Apple iPhone 6 16GB Silver   
1199     APP0829                    Apple iPhone 6 Plus 16GB Silver   
1200     APP0822                     Apple iPhone 6 16GB Space Gray   
1201     APP0825                     Apple iPhone 6 64GB Space Gray   
1202     APP0826                         Apple iPhone 6 64GB Silver   
1203     APP0828                Apple iPhone 6 Plus 16GB Space Gray   
1280     APP0856                          Apple iPhone 6 128GB Gold   
1281     APP0857                Apple iPhone 6 Plus 64GB Space Gray   
1282     APP0858                    Apple iPhone 6 Plus 64GB Silver   
1283     APP0860               Apple iPhone 6 Plus 128GB Space Gray   
1759     PAC0602               Pack Synology DS1815 + + 40TB WD Red   
1899     SAN0035       SanDisk Ultra 16GB SDHC Memory Card Class 10   
3016     APP1166                        Apple iPhone 6S 16GB Silver   
3055     ELG0014                           Elgato Video Capture Mac   
3090     APP1185                Apple iPhone 6S Plus 64GB Rose Gold   
3106     APP1167                    Apple iPhone 6S 16GB Space Gray   
3107     APP1169                     16GB Apple iPhone 6S Rose Gold   
3108     APP1170                        Apple iPhone 6S 64GB Silver   
3110     APP1171                    Apple iPhone 6S 64GB Space Gray   
3112     APP1172                          Apple iPhone 6S 64GB Gold   
3114     APP1173                     64GB Apple iPhone 6S Rose Gold   
3116     APP1178                   Apple iPhone 6S Plus 16GB Silver   
3117     APP1179               Apple iPhone 6S Plus 16GB Space Gray   
3118     APP1180                     Apple iPhone 6S Plus 16GB Gold   
3119     APP1181                Apple iPhone 6S Plus 16GB Rose Gold   
3120     APP1182                   Apple iPhone 6S Plus 64GB Silver   
3121     APP1183               Apple iPhone 6S Plus 64GB Space Gray   
3122     APP1184                     Apple iPhone 6S Plus 64GB Gold   
3123     APP1186                  Apple iPhone 6s Plus 128GB Silver   
3124     APP1187              Apple iPhone 6s Plus 128GB Space Gray   
3125     APP1188                    Apple iPhone 6s Plus 128GB Gold   
3126     APP1189               Apple iPhone 6s Plus 128GB Rose Gold   
3138     APP1168                          Apple iPhone 6S 16GB Gold   
10864    APP1202                    Apple iPad Wi-Fi 32GB Pro Plata   
11766    APP1478                           Apple iPhone 16GB Silver   
11767    APP1479                             Apple iPhone 16GB Gold   
11768    APP1480                        Apple iPhone 16GB Rose Gold   
11770    APP1482                           Apple iPhone 64GB Silver   
11771    APP1483                             Apple iPhone 64GB Gold   
15932    PAC2056  Second hand - Apple Mac mini Core 2 Duo 226Ghz...   
18741  GTE0093-A  Open - G-Technology G-DRIVE HDD 7200rpm 4TB Th...   
18822    DOD0010   Dodocool Sports Wireless Stereo Headphones Black   
18838    CAL0010  Station 3 Caldigit Thunderbolt Thunderbolt out...   
19092  SYN0173-A      Open - Synology DS218 + NAS server Mac and PC   
19185    RIN0017                                    Chime Bell Ring   

                                                    desc      price  \
307                         Pendrive USB 2.0 Mac and PC.          8   
530             USB 3.0 flash drive 16GB USB Mac and PC.         22   
792               New iPhone 5S 16G Libre (ME432Y / AB).  4.694.994   
797                New Free iPhone 5S 16GB (ME433Y / A).  4.090.042   
798                New Free iPhone 5

####missing value in desc

looking for missing values in **desc**, pattern found: majority has *open* in **name** and **promo_price** consists two decimal point values

In [76]:
products_df[products_df.desc.isna()]

sku                                               name desc  \
16126  WDT0211-A  Open - Purple 2TB WD 35 "PC Security Mac hard ...  NaN   
16128  APP1622-A  Open - Apple Smart Keyboard Pro Keyboard Folio...  NaN   
17843    PAC2334             Synology DS718 + NAS Server | 10GB RAM  NaN   
18152  KAN0034-A  Open - Kanex USB-C Gigabit Ethernet Adapter Ma...  NaN   
18490    HTE0025  Hyper Pearl 1600mAh battery Mini USB Mirror an...  NaN   
18612    OTT0200    OtterBox External Battery Power Pack 20000 mAHr  NaN   
18690  HOW0001-A  Open - Honeywell thermostat Lyric zonificador ...  NaN   

           price promo_price  in_stock      type  
16126        107     814.659         0      1298  
16128  1.568.206   1.568.206         0      1298  
17843     566.35   5.659.896         0  12175397  
18152      29.99     237.925         0      1298  
18490      24.99       22.99         1      1515  
18612      79.99       56.99         1      1515  
18690     199.99   1.441.174         0  11905404

###Deal with missing values


In [86]:
#desc column
products_df.desc.fillna(value='NO description', inplace=True)

filles the missing value with *no description*

In [91]:
products_df.loc[products_df['sku']=='PAC2334',:] #cheking if NAN value is filles

sku                                    name            desc  \
17843  PAC2334  Synology DS718 + NAS Server | 10GB RAM  NO description   

        price promo_price  in_stock      type  
17843  566.35   5.659.896         0  12175397

##price


In [97]:
#price column
products_df.price.dropna(inplace=True)

In [108]:
products_df.price.isna().value_counts(normalize=True)

False    0.995652
True     0.004348
Name: price, dtype: float64



not taking the missing value in price column


In [134]:
products_df = products_df.loc[~products_df.price.isna(), :]

In [135]:
products_df.shape

(10157, 7)

### Check / Change Data types

In [100]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10580 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10580 non-null  object
 1   name         10580 non-null  object
 2   desc         10580 non-null  object
 3   price        10534 non-null  object
 4   promo_price  10580 non-null  object
 5   in_stock     10580 non-null  int64 
 6   type         10530 non-null  object
dtypes: int64(1), object(6)
memory usage: 661.2+ KB


*price* column is object data type not float, so attempted to make it in float data type using to_numeric

In [112]:
products_df["price"] = pd.to_numeric(products_df["price"])

ValueError: Unable to parse string "1.639.792" at position 504

didnot work because it has anomaly like 1.639.792

In [113]:
products_df.price.str.contains("\d+\.\d+\.\d+").value_counts()

False    10157
True       377
Name: price, dtype: int64

377 rows are affected with problem

In [114]:
two_dot_percentage_1 = ((products_df.price.str.contains("\d+\.\d+\.\d+").value_counts()[1] / products_df.shape[0])*100).round(2)
print(f"The 2 dot problem represents {two_dot_percentage_1}% of the rows in our product DataFrame")

The 2 dot problem represents 3.58% of the rows in our product DataFrame


In [123]:
two_dot_sku_list = products_df.loc[products_df.price.str.contains("\d+\.\d+\.\d+"), "sku"]

products_df=products_df.loc[~products_df.sku.isin(two_dot_sku_list)]


not taking. this two dot decimals value in the updated product dataframe

In [124]:
products_df.shape

(10157, 7)

converting price into numeric data

In [125]:
products_df["price"] = pd.to_numeric(products_df["price"])

In [128]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10157 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   sku          10157 non-null  object 
 1   name         10157 non-null  object 
 2   desc         10157 non-null  object 
 3   price        10157 non-null  float64
 4   promo_price  10157 non-null  object 
 5   in_stock     10157 non-null  int64  
 6   type         10110 non-null  object 
dtypes: float64(1), int64(1), object(5)
memory usage: 634.8+ KB


changing data type of *promo_price* into float

In [129]:
products_df["promo_price"] = pd.to_numeric(products_df["promo_price"])

ValueError: Unable to parse string "1.119.976" at position 49

In [130]:
products_df.promo_price.str.contains("\d+\.\d+\.\d+").value_counts()

False    5836
True     4321
Name: promo_price, dtype: int64

4321 rows are affected

In [131]:
two_dot_percentage_2 = ((products_df.promo_price.str.contains("\d+\.\d+\.\d+").value_counts()[1] / products_df.shape[0])*100).round(2)
print(f"The 2 dot problem represents {two_dot_percentage_2}% of the rows in our product DataFrame")

The 2 dot problem represents 42.54% of the rows in our product DataFrame


it would be triclky if we remove this huge amount of data in total dataframe